### PR TITLE
WIP: Support for save and save as

### PR DESCRIPTION
**Pull request by kethinov**
Closes https://github.com/glyphr-studio/Glyphr-Studio-Desktop/issues/21

Closes https://github.com/glyphr-studio/Glyphr-Studio-Desktop/issues/25

Closes https://github.com/glyphr-studio/Glyphr-Studio-Desktop/issues/26

Probably closes https://github.com/glyphr-studio/Glyphr-Studio-Desktop/issues/14

Makes progress on #4

### CODE REVIEW

**Comment by kethinov** [comment]
Depends on https://github.com/glyphr-studio/Glyphr-Studio-1/pull/263

**Comment by kethinov** [comment]
@mattlag and @Autre31415 this is a super rough implementation so lemme know what sucks and I'll fix it.

One thing I don't like about it already is it only works via keyboard commands (ctrl/cmd+s for save and ctrl/cmd+shift+s for save as). The visual save button is not handled.

Open to ideas for improvements on how it should work.

**Comment by mattlag** [comment]
I agree, this needs a bit more thought.

1. Handle it via the (Electron-only) File > Save As option.  File menu doesn't exist in the browser version, so could be surfaced / handled separately here.

2. Treat the save button in the left bar as "Save As" *if electron*, but *if browser* then treat as current save experience.

**Comment by kethinov** [comment]
Okay yeah that all makes sense. I'll push an update with those changes.

Can you publish a [new version](https://github.com/glyphr-studio/Glyphr-Studio-1/releases) of the main repo that has my changes to it so we can inherit it in the desktop app?

**Comment by kethinov** [comment]
All right, pushed up some changes that reflect those additional requirements.

Pretty hacky in how it interacts with that save button, but it seems to work. If we want to clean up the interactions between how electron talks to that save button, I'll need to PR the main project again. Up to you which way you think we should go with it.

File: renderer.js [hunk 1]
```diff
@@ -45,16 +46,52 @@ saveFile = function (fname, buffer, ftype) { // eslint-disable-line
...
-      fs.writeFile(destination + '/' + fname, buffer)
+    if (window.saveFileOverwrite && window.saveFileOverwriteFile) {
+      fs.writeFileSync(window.saveFileOverwriteFile, buffer)
+      alert('Saved to ' + window.saveFileOverwriteFile)
     } else {
-      event.returnValue = 'false'
+      dialog.showSaveDialog({
+        properties: ['openFile'],
+        title: 'Choose where to save project...',
+        defaultPath: process.env.HOME + '/' + fname
```
Comment (Autre31415) [inline renderer.js]: Use path.join for the path concat

File: renderer.js [hunk 2]
```diff
@@ -45,16 +46,52 @@ saveFile = function (fname, buffer, ftype) { // eslint-disable-line
...
-      defaultPath: process.env.HOME
-    })
-
-    if (destination !== undefined) {
-      fs.writeFile(destination + '/' + fname, buffer)
+    if (window.saveFileOverwrite && window.saveFileOverwriteFile) {
+      fs.writeFileSync(window.saveFileOverwriteFile, buffer)
+      alert('Saved to ' + window.saveFileOverwriteFile)
     } else {
-      event.returnValue = 'false'
```
Comment (Autre31415) [inline renderer.js]: Without a replacement for this line, selecting "Cancel" on the "Would you like to save before closing?" dialog will always cause the app to close.

Comment (kethinov) [reply renderer.js]: done

File: renderer.js (hunk 1 above)
Comment (kethinov) [reply renderer.js]: done

File: renderer.js (hunk 2 above)
Comment (kethinov) [reply renderer.js]: fixed the bug associated with this

**Comment by kethinov** [comment]
I suppose one more thing I should do on this at least before we merge it is suppress the save/save-as menus on the welcome screen.

**Comment by kethinov** [comment]
Okay, took care of suppressing the save/save-as menus on the welcome screen.

**Comment by kethinov** [comment]
@mattlag and @Autre31415 I think I'm done tweaking it now. Give it a spin and let me know if you think this is good now.

**Comment by Autre31415** [comment]
@mattlag I've issued a release that includes these changes [here](https://github.com/glyphr-studio/Glyphr-Studio-Desktop/releases/tag/v0.4.2). If you think any other changes should be made let us know and we'll put up another PR.

**Pull request merged**